File: cpp/Linked_List/0206_Reverse_Linked_List.cpp
using namespace std;

struct ListNode {
  int val;
  ListNode *next;

  ListNode() : val(0), next(nullptr) {}

  ListNode(int x) : val(x), next(nullptr) {}

  ListNode(int x, ListNode *next) : val(x), next(next) {}
};
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  /*
   * 1st solution recursive - reverse from back
   * */
  ListNode *reverseList(ListNode *head) {
    if (!head) return nullptr;
    if (!head->next) return head;
    ListNode *last = reverseList(head->next);
    head->next->next = head;
    head->next = nullptr;
    return last;
  }

  /*
   * 2nd solution two pointers
   * */
  ListNode *reverseList2(ListNode *head) {
    ListNode *pre = nullptr, *cur = head, *temp = nullptr;
    while (cur) {
      temp = cur->next;
      cur->next = pre;
      pre = cur;
      cur = temp;
    }
    return pre;
  }

  /*
   * 3rd solution recursive (based on two pointers)
   * */
  ListNode *reverseList3(ListNode *head) {
    return reverse(nullptr, head);
  }

  ListNode *reverse(ListNode *pre, ListNode *cur) {
    if (!cur) return pre;
    ListNode *temp = cur->next;
    cur->next = pre;
    return reverse(cur, temp);
  }
};
//leetcode submit region end(Prohibit modification and deletion)

File: python3/0206-reverse-linked-list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    # iterative
    def reverseList(self, head: [ListNode]) -> [ListNode]:
        cur = head
        pre = None
        while cur:
            temp = cur.next
            cur.next = pre
            pre = cur
            cur = temp
        return pre

    #recursive
    def reverseList2(self, head: [ListNode]) -> [ListNode]:
        if not head or not head.next:
            return head
        else:
            last_node = self.reverseList2(head.next)
            head.next.next = head
            head.next = None
        return last_node

def to_linked_list(iterable):
    head = None
    for val in reversed(iterable):
        head = ListNode(val, head)
    return head


def to_native_list(head):
    lst = []
    while head:
        lst.append(head.val)
        head = head.next
    return lst


head = [1, 2, 3, 4, 5]
l1 = to_linked_list(head)
result = Solution().reverseList(l1)
print(to_native_list(result))

head = [1, 2]
l2 = to_linked_list(head)
result = Solution().reverseList(l2)
print(to_native_list(result))
